### PR TITLE
Reverted fetch-cobolcheck and fetch-cobolcheck.ps1 scripts.

### PR DESCRIPTION
**Pull request by SimaDovakin**
Attempt to fix the problem with cobolcheck fetching.

### CODE REVIEW

**Comment by kapitaali** [comment]
what say @ErikSchierboom ?

**Review by ErikSchierboom — COMMENTED** [review]
Could you explain why these changes would fix things? It looks to me like they no longer use a shared directory and thus would re-download cobolcheck for each exercise.

**Comment by SimaDovakin** [comment]
It required root permissions to download cobolcheck to `/usr/local/bin`, I guess. Maybe we can download cobolcheck to the repo's bin directory and then check if it is already here.

**Comment by ErikSchierboom** [comment]
> Maybe we can download cobolcheck to the repo's bin directory and then check if it is already here.

Yes, either that or store it in `$HOME/cobol` or `$HOME/.cobol` or something.

And then update the `test.sh` scripts.

**Comment by kapitaali** [comment]
Per the old fetch-cobolcheck source, it fetches it to each repo's ./bin. This PR basically reverts back to the old functionality by adding

```
  if [[ -d ./bin ]]; then
    output_dir="./bin"
  elif [[ $PWD == */bin ]]; then
    output_dir="$PWD"
  else
    echo "Error: no ./bin directory found. This script should be ran from a repo root." >&2
    return 1
  fi

```

**Comment by SimaDovakin** [comment]
> > Maybe we can download cobolcheck to the repo's bin directory and then check if it is already here.
> 
> Yes, either that or store it in `$HOME/cobol` or `$HOME/.cobol` or something.
> 
> And then update the `test.sh` scripts.

It makes sense. We can try to fetch cobolcheck and download it to `$HOME/cobolcheck`. I'll try to do it.

**Comment by ErikSchierboom** [comment]
Using the repo root is fine too. I don't really care as long as one place is used

**Comment by SimaDovakin** [comment]
@kapitaali @ErikSchierboom Now cobolcheck is stored in `$HOME/cobolcheck/cobolcheck{,exe}`. I renamed `bin/sync-fetch-cobolcheck-files` to `bin/sync-exercise-files` and added synchronizations for `test.sh` and `test.ps1`.

File: bin/fetch-cobolcheck.ps1 [hunk 1]
```diff
@@ -23,5 +23,5 @@ Function Get-DownloadUrl {
 }
 
 $downloadUrl = Get-DownloadUrl
-$outputFile = Join-Path -Path $PSScriptRoot -ChildPath "cobolcheck.exe"
+$outputFile = Join-Path -Path "${HOME}" "cobolcheck" "cobolcheck.exe"
```
Comment (ErikSchierboom) [inline bin/fetch-cobolcheck.ps1:26]: Does `Invoke-WebRequest` create the `cobolcheck` directory when it doesn't yet exist?

Comment (SimaDovakin) [reply bin/fetch-cobolcheck.ps1]: I guess, no. Didn't test it. I added additional `New-Item` command to create it now.

**Review by ErikSchierboom — APPROVED** [review]
Let's try this out

**Comment by SimaDovakin** [comment]
@kapitaali @ErikSchierboom could we merge it?

**Comment by ErikSchierboom** [comment]
Done!

**Comment by kapitaali** [comment]
The tests should work out of the box if you download the exercise with cli. I deleted the anagram directory, downloaded it with cli, copied my source code to src, ran test.sh and got the following:

```
exercism/cobol/anagram 
❯ sh test.sh
test.sh: 2: Bad substitution
test.sh: 7: [[: not found
Cobolcheck not found, try to fetch it.
curl: (56) Failure writing output to destination, passed 1369 returned 4294967295
test.sh: 16: ./bin/cobolcheck: not found
COMPILE AND RUN TEST
cobc: test.cob: No such file or directory
```

**Comment by SimaDovakin** [comment]
@kapitaali Did you try `./test.sh` instead of `sh test.sh`?

**Comment by kapitaali** [comment]
It should make no difference, but trying ./test.sh gives me

```
exercism/cobol/anagram 
|➜ ./test.sh
bash: ./test.sh: Permission denied
```

**Comment by SimaDovakin** [comment]
I got `Permission denied` too but after `chmod +x test.sh` it works.

**Comment by kapitaali** [comment]
Doing chmod +x test.sh allows me to execute it by running ./test.sh, but the errors persist:

```
exercism/cobol/anagram 
|➜ chmod +x test.sh 

exercism/cobol/anagram 
|➜ ./test.sh 
Cobolcheck not found, try to fetch it.
curl: (56) Failure writing output to destination, passed 1369 returned 4294967295
./test.sh: line 16: ./bin/cobolcheck: No such file or directory
COMPILE AND RUN TEST
cobc: test.cob: No such file or directory
```

**Comment by SimaDovakin** [comment]
Exercism COBOL documentation says that you have to run `bash test.sh` for running tests https://exercism.org/docs/tracks/cobol/tests. Actually it does make differences because `sh` and `bash` are a little different now.

**Comment by SimaDovakin** [comment]
> Doing chmod +x test.sh allows me to execute it by running ./test.sh, but the errors persist:
> 
> ```
> exercism/cobol/anagram 
> |➜ chmod +x test.sh 
> 
> exercism/cobol/anagram 
> |➜ ./test.sh 
> Cobolcheck not found, try to fetch it.
> curl: (56) Failure writing output to destination, passed 1369 returned 4294967295
> ./test.sh: line 16: ./bin/cobolcheck: No such file or directory
> COMPILE AND RUN TEST
> cobc: test.cob: No such file or directory
> ```

Hmm.. It's weird. It works for me. What's your OS?

**Comment by kapitaali** [comment]
This is Pop!_OS, an Ubuntu variant. 

I am just wondering why would we need to touch Linux shell scripts at all if the original intention was to do something about Powershell.

**Comment by SimaDovakin** [comment]
We changed `test.sh` too to be able to fetch cobolcheck only once if it doesn't already exist in the home directory. I use Linux Mint and it works for me

**Comment by kapitaali** [comment]
We now have 1 datapoint that works and 1 that doesn't. Need more. I have a clean debian install, going to try with that. Just a sec.

**Comment by kapitaali** [comment]
Nope, it fails on debian:

```
~/exercism/cobol/anagram$ ./test.sh
bash: ./test.sh: Permission denied
~/exercism/cobol/anagram$ chmod +x test.sh
~/exercism/cobol/anagram$ ./test.sh
Cobolcheck not found, try to fetch it.
curl: (23) Failure writing output to destination
./test.sh: line 16: ./bin/cobolcheck: No such file or directory
COMPILE AND RUN TEST
./test.sh: line 20: cobc: command not found
~/exercism/cobol/anagram$
```

**Comment by kapitaali** [comment]
Fails on debian too:

```
~/exercism/cobol/anagram$ ./test.sh
Cobolcheck not found, try to fetch it.
curl: (23) Failure writing output to destination
./test.sh: line 16: ./bin/cobolcheck: No such file or directory
COMPILE AND RUN TEST
cobc: test.cob: No such file or directory
```

**Comment by SimaDovakin** [comment]
It's really weird. It seems like you don't have permissions to write to home directory. This is my try:
[asciicast](https://asciinema.org/a/XVS5hCnWwqPG6tmqT0FKLL5D4)

**Comment by kapitaali** [comment]
I believe it works with your system, but running ./test.sh should work everywhere and be configuration independent. 

Not a single user should have to make any configurations to their system to run test.sh and COBOL tests succesfully. That was how the testing script worked a while ago.

**Comment by SimaDovakin** [comment]
Can you share your `test.sh` please? I this case, maybe, we really should revert changes.

**Comment by kapitaali** [comment]
Yeah it's here:

```
#!/usr/bin/env bash
SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )/
SLUG=${1:-$(basename "${SCRIPT_DIR}")}
COBOLCHECK=./bin/cobolcheck

WHICH_COBOLCHECK=$(which cobolcheck)
if [[ $? -eq 0 ]] ; then
    echo "Found cobolcheck, using $COBOLCHECK"
    COBOLCHECK=$WHICH_COBOLCHECK
elif [ ! -f $SCRIPT_DIR/bin/cobolcheck ]; then
    echo "Cobolcheck not found, try to fetch it."
    cd $SCRIPT_DIR/bin/
    bash fetch-cobolcheck
fi
cd $SCRIPT_DIR
$COBOLCHECK -p $SLUG

# compile and run
echo "COMPILE AND RUN TEST"
cobc -xj test.cob
```

**Comment by SimaDovakin** [comment]
This is an old version of `test.sh`. There is new one on the main branch.

**Comment by ErikSchierboom** [comment]
@SimaDovakin Could you look at this: https://forum.exercism.org/t/test-setup-failure-through-online-editor-for-cobol-exercise-secret-handshake/12710

**Comment by SimaDovakin** [comment]
@ErikSchierboom sure. It is weird. It works locally but it can't resolve api.github.com.

**Comment by SimaDovakin** [comment]
> @SimaDovakin Could you look at this: https://forum.exercism.org/t/test-setup-failure-through-online-editor-for-cobol-exercise-secret-handshake/12710

@kapitaali @ErikSchierboom Could it be connected with DNS resolving on the server? And why?

**Comment by kapitaali** [comment]
The fetch-cobolcheck connects to this url: https://api.github.com/repos/0xE282B0/cobol-check/releases/latest

It seems that the url in question returns a JSON response, which the fetch-cobolcheck script does not handle. Better change the url to the one in the response: https://github.com/0xE282B0/cobol-check/releases/download/v0.2.0-RC1-graal/cobol-check-linux-aarch64

**Comment by SimaDovakin** [comment]
By the errors, it seems that curl try to resolve api.github.com but it can't for some reason.

**Comment by kapitaali** [comment]
I deleted cobolcheck and ran test.sh and fetch-cobolcheck worked fine here too. So it's the test environment that is failing.

**Comment by kapitaali** [comment]
I ran tests on the website and they're working. So it must have been a temporary hiccup that is now fixed.

**Comment by SimaDovakin** [comment]
Hmm.. interesting

**Comment by kapitaali** [comment]
The exercise that I tried that worked was anagram, the one that was added last. But updating the atbash-cipher exercise on exercism.org gave me the error that was reported:

```
Cobolcheck not found, try to fetch it.
curl: (6) Could not resolve host: api.github.com
curl: (6) Could not resolve host: api.github.com
curl: (6) Could not resolve host: api.github.com
curl: (6) Could not resolve host: api.github.com
curl: (3) URL using bad/illegal format or missing URL
chmod: cannot access '/root/cobolcheck/cobolcheck': No such file or directory
Cobolcheck has been downloaded to /root/cobolcheck/cobolcheck
/mnt/exercism-iteration/test.sh: line 12: /root/cobolcheck/cobolcheck: No such file or directory
COMPILE AND RUN TEST
cobc: test.cob: No such file or directory  
SOMETHING WENT WRONG DURING TEST SETUP, PLEASE OPEN A TICKET AT: https://github.com/exercism/cobol/issues/new
```

I'm going to try downloading an updated exercise with the cli to see if this happens locally.

**Comment by kapitaali** [comment]
Running locally the updated exercise that has been downloaded with cli works fine. cobolcheck gets downloaded and tests are run.

**Comment by ErikSchierboom** [comment]
The test runner can't download anything as it does not have internet. Cobolcheck thus must be downloaded to the right location in the test runner dockerfile. Can you check that?

**Comment by SimaDovakin** [comment]
I checked that locally earlier and it works fine. I ran test runner locally in container and without it.

**Comment by ErikSchierboom** [comment]
What command did you use?

**Comment by kapitaali** [comment]
There are things that have been paused in the workflow: https://github.com/exercism/cobol/actions/runs/10566009241/workflow

**Comment by SimaDovakin** [comment]
> What command did you use?

As I remember I used `./bin/run.sh ...` and `./bin/run-in-docker.sh ...`.

File: exercises/practice/acronym/test.sh [hunk 2]
```diff
@@ -1,13 +1,9 @@
...
 SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )/
 SLUG=${1:-$(basename "${SCRIPT_DIR}")}
-COBOLCHECK=./bin/cobolcheck
+COBOLCHECK="${HOME}/cobolcheck/cobolcheck"
 
-WHICH_COBOLCHECK=$(which cobolcheck)
-if [[ $? -eq 0 ]] ; then
-    echo "Found cobolcheck, using $COBOLCHECK"
-    COBOLCHECK=$WHICH_COBOLCHECK
-elif [ ! -f $SCRIPT_DIR/bin/cobolcheck ]; then
```
Comment (ErikSchierboom) [inline exercises/practice/acronym/test.sh:10]: @SimaDovakin I think the remove of this part is what breaks the test runner. The cobol test runner downloads the cobolcheck file to: https://github.com/exercism/cobol-test-runner/blob/main/Dockerfile#L16 

What I would suggest to do is:

- Either revert the WHICH part that was deleted, or
- Change the location of cobolcheck in the test runner (and also update the `test.sh` files in the `tests/*` directories to match what is used here.

Comment (SimaDovakin) [reply exercises/practice/acronym/test.sh:10]: Hmm... Ok, I'll return "WHICH" part back.

**Comment by SimaDovakin** [comment]
@ErikSchierboom @kapitaali  I reverted WHICH part here https://github.com/exercism/cobol/pull/162. But I'm still wondering why the test runner can't resolve `api.github.com`. Is it because it doesn't have the appropriate DNS settings inside the container?

**Pull request merged**